File: packages/shared/src/types/preference.ts
/**
 * UserPreference（用户偏好）类型定义
 *
 * 对应 Prisma model UserPreference（user_preferences 表）。
 * 列式存储（非 JSON），Prisma 类型安全、可索引（SET-P0-02）。
 *
 * 偏好设置跟随账号，实现多端/多浏览器一致。
 * 首次登录返回系统默认值，保存为乐观更新 + 失败回滚。
 */

/**
 * 默认粒度枚举
 */
export const PreferenceGranularity = {
  DAY: 'day',
  WEEK: 'week',
  MONTH: 'month',
  YEAR: 'year',
} as const;

export type PreferenceGranularity =
  (typeof PreferenceGranularity)[keyof typeof PreferenceGranularity];

/**
 * 默认日期范围枚举
 */
export const PreferenceDateRange = {
  /** 近 3 月 */
  THREE_MONTHS: '3m',
  /** 近 1 年 */
  ONE_YEAR: '1y',
  /** 今年至今 */
  YEAR_TO_DATE: 'ytd',
  /** 全部 */
  ALL: 'all',
} as const;

export type PreferenceDateRange =
  (typeof PreferenceDateRange)[keyof typeof PreferenceDateRange];

/**
 * 聚合方式枚举（复用 QueryGranularity 中的概念，这里是偏好存储的值）
 */
export const PreferenceAggregation = {
  /** 取期末值 */
  LAST: 'last',
  /** 取平均值 */
  AVG: 'avg',
} as const;

export type PreferenceAggregation =
  (typeof PreferenceAggregation)[keyof typeof PreferenceAggregation];

/**
 * 主题枚举
 */
export const PreferenceTheme = {
  LIGHT: 'light',
  DARK: 'dark',
  SYSTEM: 'system',
} as const;

export type PreferenceTheme =
  (typeof PreferenceTheme)[keyof typeof PreferenceTheme];

/**
 * 用户偏好实体
 */
export interface UserPreference {
  /** UUID 主键 */
  id: string;
  /** 用户 ID（唯一） */
  userId: string;
  /** 登录后自动选中的组合 ID */
  defaultPortfolioId: string | null;
  /** 默认时间维度，默认 "month" */
  defaultGranularity: string;
  /** 默认日期范围快捷项，默认 "1y" */
  defaultDateRange: string;
  /** 周期聚合方式，默认 "last" */
  aggregation: string;
  /** 按周聚合的周起始日（0=周日, 1=周一），默认 1 */
  weekStartsOn: number;
  /** 净值展示小数位，默认 4 */
  navDecimals: number;
  /** XIRR 百分比小数位，默认 2 */
  xirrDecimals: number;
  /** 主题（light/dark/system），默认 "system" */
  theme: string;
  /** 快照过期提醒阈值（天数），默认 3 */
  staleDays: number;
  /** 创建时间 ISO 8601 */
  createdAt: string;
  /** 更新时间 ISO 8601 */
  updatedAt: string;
}

/**
 * 更新偏好 DTO（所有字段可选）
 */
export interface UpdatePreferenceDto {
  /** 登录后自动选中的组合 ID */
  defaultPortfolioId?: string | null;
  /** 默认时间维度 */
  defaultGranularity?: string;
  /** 默认日期范围快捷项 */
  defaultDateRange?: string;
  /** 周期聚合方式 */
  aggregation?: string;
  /** 按周聚合的周起始日 */
  weekStartsOn?: number;
  /** 净值展示小数位 */
  navDecimals?: number;
  /** XIRR 百分比小数位 */
  xirrDecimals?: number;
  /** 主题 */
  theme?: string;
  /** 快照过期提醒阈值 */
  staleDays?: number;
}

File: backend/app/routers/preference.py
"""用户偏好路由（§4.2.16 · SET-P0-02）。

GET /api/users/preferences  → 取（不存在则建默认）偏好
PATCH /api/users/preferences → 全站唯一偏好写入口（部分更新 + 服务端白名单校验）
"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.enums import BusinessErrorCode
from app.core.envelope import EnvelopeRoute
from app.core.exceptions import BusinessException
from app.core.security import CurrentUser, get_current_user
from app.db.database import get_db
from app.models import UserPreference
from app.routers.common import serialize_preference


# 服务端白名单（裁决 Q-5：保持 String + 校验，零 migration）
_DATE_RANGES = ["1w", "1m", "3m", "6m", "1y", "ytd", "all"]
_GRANULARITIES = ["day", "week", "month", "quarter", "year"]
_AGGREGATIONS = ["last", "avg"]
_THEMES = ["system", "light", "dark"]

ROUTER_PREFIX = "/api/users"


router = APIRouter(prefix=ROUTER_PREFIX, tags=["preference"], route_class=EnvelopeRoute)


async def _get_or_create(db: AsyncSession, user_id: str) -> UserPreference:
    pref = (
        await db.execute(
            select(UserPreference).where(UserPreference.user_id == user_id)
        )
    ).scalar_one_or_none()
    if pref is None:
        pref = UserPreference(user_id=user_id)
        db.add(pref)
        await db.commit()
        await db.refresh(pref)
    return pref


def _validate(field: str, value: str, allowed: list[str]) -> None:
    if value not in allowed:
        raise BusinessException(
            code=BusinessErrorCode.VALIDATION_FAILED,
            message=f"{field} 取值无效：{value}（允许 {', '.join(allowed)}）",
            status_code=400,
        )


@router.get("/preferences")
async def get_preferences(
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    pref = await _get_or_create(db, user.user_id)
    return serialize_preference(pref)


@router.patch("/preferences")
async def patch_preferences(
    body: dict,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    pref = await _get_or_create(db, user.user_id)

    # 白名单字段（防止任意列注入）
    allowed = {
        "defaultPortfolioId",
        "defaultGranularity",
        "defaultDateRange",
        "aggregation",
        "weekStartsOn",
        "navDecimals",
        "xirrDecimals",
        "theme",
        "staleDays",
        "showLiquidated",
        "costBasisView",
        "cashHintOnCashflow",
        "cashHintOnTrade",
        "amountThousands",
        "amountAbbrev",
    }
    unknown = set(body.keys()) - allowed
    if unknown:
        raise BusinessException(
            code=BusinessErrorCode.VALIDATION_FAILED,
            message=f"未知偏好字段：{', '.join(sorted(unknown))}",
            status_code=400,
        )

    for key, value in body.items():
        if value is None:
            continue
        if key == "defaultDateRange":
            _validate("defaultDateRange", value, _DATE_RANGES)
        elif key == "defaultGranularity":
            _validate("defaultGranularity", value, _GRANULARITIES)
        elif key == "aggregation":
            _validate("aggregation", value, _AGGREGATIONS)
        elif key == "theme":
            _validate("theme", value, _THEMES)
        # 其余字段按类型直接赋值（由 ORM/DB 约束保证合法）
        setattr(pref, _snake(key), value)

    await db.commit()
    await db.refresh(pref)
    return serialize_preference(pref)


def _snake(camel: str) -> str:
    out = []
    for ch in camel:
        if ch.isupper():
            out.append("_")
            out.append(ch.lower())
        else:
            out.append(ch)
    return "".join(out)
